src/sinks/kafka/service.rs

```diff
@@ -0,0 +1,115 @@
+use crate::buffers::Ackable;
+use crate::event::{EventFinalizers, EventStatus, Finalizable};
+use crate::kafka::KafkaStatisticsContext;
+use crate::sinks::kafka::config::QUEUED_MIN_MESSAGES;
+use futures::future::BoxFuture;
+use rdkafka::error::KafkaError;
+use rdkafka::message::OwnedHeaders;
+use rdkafka::producer::{FutureProducer, FutureRecord};
+use rdkafka::util::Timeout;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::time::Duration;
+use tokio::time::Sleep;
+use tower::Service;
+
+pub struct KafkaRequest {
+    pub body: Vec<u8>,
+    pub metadata: KafkaRequestMetadata,
+}
+
+pub struct KafkaRequestMetadata {
+    pub finalizers: EventFinalizers,
+    pub key: Option<Vec<u8>>,
+    pub timestamp_millis: Option<i64>,
+    pub headers: Option<OwnedHeaders>,
+    pub topic: String,
+}
+
+pub struct KafkaResponse {}
+
+impl AsRef<EventStatus> for KafkaResponse {
+    fn as_ref(&self) -> &EventStatus {
+        &EventStatus::Delivered
+    }
+}
+
+impl Ackable for KafkaRequest {
+    fn ack_size(&self) -> usize {
+        // rdkafka takes care of batching internally, so a request here is always 1 event
+        1
+    }
+}
+
+impl Finalizable for KafkaRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        std::mem::take(&mut self.metadata.finalizers)
+    }
+}
+
+pub struct KafkaService {
+    kafka_producer: FutureProducer<KafkaStatisticsContext>,
+    current_in_flight: Arc<AtomicU64>,
+    delay: Option<Pin<Box<Sleep>>>,
+}
+
+impl KafkaService {
+    pub fn new(kafka_producer: FutureProducer<KafkaStatisticsContext>) -> KafkaService {
+        KafkaService {
+            kafka_producer,
+            current_in_flight: Arc::new(AtomicU64::new(0)),
+            delay: None,
+        }
+    }
+}
+
+impl Service<KafkaRequest> for KafkaService {
+    type Response = KafkaResponse;
+    type Error = KafkaError;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // rdkafka has an `in_flight_count()` but that can't be used here
+        // since it doesn't count requests that have not been queued yet
+        if self.current_in_flight.load(Ordering::Relaxed) < QUEUED_MIN_MESSAGES {
+            return Poll::Ready(Ok(()));
+        }
+
+        // This is the same amount of time that rdkafka delays when the internal queue is full
+        let mut sleep = Box::pin(tokio::time::sleep(Duration::from_millis(100)));
+        let result = sleep.as_mut().poll(cx).map(|_| Ok(()));
+        self.delay = Some(sleep); // prevent the timer from being dropped
+        result
```

Comment:
Normally we would use [`ConcurrencyLimit`](https://docs.rs/tower/0.4.8/tower/limit/concurrency/struct.ConcurrencyLimit.html) for limiting concurrency of a `Service`. You can look at the docs for [`ServiceBuilder`](https://docs.rs/tower/0.4.8/tower/struct.ServiceBuilder.html)

As you have it now, not only do we have this extraneous atomic, but you're also going to wake up the task every 100ms regardless of whether or not an in-flight request has actually completed.

Reply:
yep, that seems a lot nicer. I'll use `ConcurrencyLimit` instead

Reply:
changed in 72c93aa06e